### PR TITLE
eth/protocols, prp/tracker: add support for req/rep rtt tracking

### PR DESCRIPTION
**Pull request by karalabe**
This PR is a counterpart to https://github.com/ethereum/go-ethereum/pull/22581, adding metrics to the request/reply RTTs too. 

Given that we (currently) do not have a centralized dispatcher which matches requests with replies, this PR introduces a global request tracker (per protocol). The tracker can be notified of outbound requests and inbound replies, and it will match them up and update the relevant metrics.

Since the request/reply pattern is stateful, there are a few caveats in the tracker:

- The tracker maintains a list of all currently pending requests. The only way to remove an item from this list is to either deliver a response; or for the request to time out. The timeout however is not the same as the `downloader` might use, rather a global number per protocol (5 min for `eth`, 1 min for `snap`). This allows us to a) detect loooong deliveries even if the calling code timed out and discarded the request; b) track expiries with a doubly linked list, making deletions and expiry O(1).
- Given that we track every send request, these can accumulate quite a bit if we have a huge peer count. In theory it shouldn't go over the number of peers / packet type, but it can temporarily on sync restarts or peer churn. I've added a metric to display these too to avoid surprises, but we might consider some hard enforced limit to make sure we can't go OOM due to some bug.

Open questions:

- I've used one singleton global tracker per protocol. Creating a scoped tracker in package `eth` (backend) and propagating it to all relevant call sites would become quite nasty; and our metrics can't differentiate between 2 `node` instances running within the same process either way. May or may not be a goo idea, wanted to emphasize it.
- We could fairly easily add the possibility to track request/reply sizes too. Unsure if this would be more help or more noise.

TL;DR

Screenshot from 2021-04-01 10-25-52

### CODE REVIEW

File: p2p/tracker/tracker.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+func (t *Tracker) schedule() {
+	if t.expire.Len() == 0 {
+		t.wake = nil
+		return
+	}
+	t.wake = time.AfterFunc(time.Until(t.pending[t.expire.Front().Value.(uint64)].time.Add(t.timeout)), t.clean)
+}
+
+// Fulfil fills a pending request, if any is available, reporting on various metrics.
+func (t *Tracker) Fulfil(peer string, version uint, code uint64, id uint64) {
```
Comment (holiman) [inline p2p/tracker/tracker.go:155]:
```suggestion
func (t *Tracker) Fulfill(peer string, version uint, code uint64, id uint64) {
```

**Review by holiman — APPROVED** [review]
LGTM, if you add some kind of hard limit to the number of tracked requests (it can be large though)

**Comment by karalabe** [comment]
@holiman PTAL

Added a cap, looks like this when hit (I set it to 1 to test it):

Screenshot from 2021-04-21 14-20-16
